### PR TITLE
Refactor dense network code into its own class with config

### PR DESCRIPTION
**Pull request by mcgibbon**
The dense and precipitative models re-use a lot of the same configuration and code. This PR removes this duplication by creating a DenseNetwork class for configuring dense sub-networks to use when building a larger network.

Added public API:
- Added OptimizerConfig and RegularizerConfig to fv3fit namespace since they are already public API
- Added new classes DenseNetworkConfig and DenseNetwork

Significant internal changes:
- moved DenseHyperparameters to `fv3fit.keras._models.models`
- refactored dense and precipitative training routines to use DenseNetworkConfig and DenseNetwork

- [x] Tests added

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/shared.py [hunk 1]
```diff
@@ -44,6 +47,79 @@ def get_input_vector(
     return input_layers, packed
 
 
+@dataclasses.dataclass
+class DenseNetwork:
```
Comment (brianhenn) [inline external/fv3fit/fv3fit/keras/_models/shared.py:51]: I like this structure for easily getting at intermediate layers

**Review by brianhenn — APPROVED** [review]
Thanks @mcgibbon I have a suggestion about the dense network but this looks good

File: external/fv3fit/fv3fit/keras/_models/models.py [hunk 2]
```diff
@@ -42,6 +46,54 @@
...
+    """
+
+    input_variables: List[str]
+    output_variables: List[str]
+    weights: Optional[Mapping[str, Union[int, float]]] = None
+    normalize_loss: bool = True
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    dense_network: DenseNetworkConfig = dataclasses.field(
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/models.py:81]: Yes.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 3]
```diff
@@ -358,10 +328,12 @@ def _build_model(self) -> Tuple[tf.keras.Model, tf.keras.Model]:
         T_tendency = unpacked_output[0]
         q_tendency = unpacked_output[1]
 
+        # share hidden layers with the dense network,
+        # but no activity regularization for column precipitation
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:332]: The output tensor returned in the return value can be ignored in a case where you want to manually construct the output layer. Here, for example, I could write both of these to act on the last hidden layer, possibly meaning I can remove the "activity regularization" option from the dense network config. How does this sound, in comparison to what's done here right now?

File: external/fv3fit/fv3fit/keras/_models/shared.py [hunk 4]
```diff
@@ -44,6 +47,79 @@ def get_input_vector(
...
+            if self.spectral_normalization:
+                hidden_layer = tfa.layers.SpectralNormalization(
+                    hidden_layer, name=f"spectral_norm_{label}_{i}"
+                )
+            x = hidden_layer(x)
+            hidden_outputs.append(x)
+        output = tf.keras.layers.Dense(
+            n_features_out,
+            activation="linear",
+            # activity_regularizer=self.output_activity_regularizer.instance,
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared.py:117]: True. I think I'll keep this as part of this function and the API, but just ignore the output tensor in the precipitative model (and continue to use it in the Dense model).

**Comment by mcgibbon** [comment]
Merging for now, once it is merged we will consider whether to make further changes (e.g. remove the output layer from densenetwork building).

**Pull request merged**